libs/langchain/langchain/storage/_lc_store.py

```diff
@@ -0,0 +1,88 @@
+"""Create a key-value store for any langchain serializable object."""
+from typing import Callable, Optional
+
+from langchain.load.dump import dumps
+from langchain.load.load import loads
+from langchain.load.serializable import Serializable
+from langchain.schema import BaseStore, Document
+from langchain.storage.encoder_backed import EncoderBackedStore
+
+
+def _dump_as_bytes(obj: Serializable) -> bytes:
+    """Return a bytes representation of a document."""
+    return dumps(obj).encode("utf-8")
+
+
+def _dump_document_as_bytes(obj: Document) -> bytes:
+    """Return a bytes representation of a document."""
+    if not isinstance(obj, Document):
+        raise TypeError("Expected a Document instance")
+    return dumps(obj).encode("utf-8")
+
+
+def _load_document_from_bytes(serialized: bytes) -> Document:
+    """Return a document from a bytes representation."""
+    obj = loads(serialized.decode("utf-8"))
+    if not isinstance(obj, Document):
+        raise TypeError(f"Expected a Document instance. Got {type(obj)}")
+    return obj
+
+
+def _load_from_bytes(serialized: bytes) -> Serializable:
+    """Return a document from a bytes representation."""
+    return loads(serialized.decode("utf-8"))
+
+
+def _identity(x: str) -> str:
+    """Return the same object."""
+    return x
+
+
+# PUBLIC API
+
+
+def create_lc_store(
+    store: BaseStore[str, bytes],
+    *,
+    key_encoder: Optional[Callable[[str], str]] = None,
+) -> BaseStore[str, Serializable]:
+    """Create a store for langchain serializable objects from a bytes store.
+
+    Args:
+        store: A bytes store to use as the underlying store.
+        key_encoder: A function to encode keys; if None uses identity function.
+
+    Returns:
+        A key-value store for documents.
+    """
+    return EncoderBackedStore(
+        store,
+        key_encoder or _identity,
+        _dump_as_bytes,
+        _load_from_bytes,
+    )
+
+
+def create_kv_docstore(
```

Comment:
Other possibility below -- but seems like over abstraction

Objects we'll want to serialize are:

1. everything (then no need to type check)
2. prompts (can do special purpose)
3. documents (can do special purpose)

----

The other possibility if anyone has time is to try and introduce a more general class that proxies the embedding class but gets the type signatures right


The tricky thing is getting the type signatures right :) 

user passes `types` and the class does run time type checking.



```


class LCStore(BaseStore[str, T]):
    def __init__(self, store: BaseStore[str, bytes], types: Sequence[Type]):
        """Create a store for langchain serializable objects from a bytes store."""
        self.types = types
        self.store = EncoderBackedStore(
            store,
            key_encoder or _identity,
            _dump_document_as_bytes,
            _load_document_from_bytes,
        )

    def mget(self, keys: Sequence[K]) -> List[Optional[V]]:
        """Get multiple keys."""
        return self.store.mget(keys)

    def mdelete(self, keys: Sequence[K]) -> None:
        """Delete multiple keys."""
        return self.store.mdelete(keys)

    def yield_keys(
        self, *, prefix: Optional[str] = None
    ) -> Union[Iterator[K], Iterator[str]]:
        """Yield all keys in the store."""
        yield from self.store.yield_keys(prefix=prefix)

    def mset(self, items: Sequence[Tuple[K, V]]) -> None:
        """Set multiple key-value pairs."""
        return self.store.mset(items)
```